src/sumo_gym/envs/fmp.py

```diff
@@ -79,58 +79,115 @@ def __init__(
 
         else:
             (
-                raw_vertices,  # id, x, y
-                # raw_charging_stations,
-                # raw_electric_vehicles,
-                raw_edges,  # edge_id, id_start, id_dest
-                raw_departures,  # vehicle_id, edge_id_from
-                raw_demand,  # start_id, dest_id
+                raw_vertices,  # [id (str), x_coord (float), y_coord (float)]
+                raw_charging_stations, # [id, (x_coord, y_coord), edge_id, charging speed]
+                raw_electric_vehicles, # [id (str), maximum speed (float), maximumBatteryCapacity (float)]
+                raw_edges,  # [id (str), from_vertex_id (str), to_vertex_id (str)]
+                raw_departures,  # [vehicle_id, starting_edge_id]
+                raw_demand,  # [junction_id, dest_vertex_id]
             ) = sumo_gym.utils.xml_utils.decode_xml_fmp(
-                net_xml_file_path, demand_xml_file_path
+                net_xml_file_path, demand_xml_file_path,
+                additional_xml_file_path
             )
 
-            # todo: better implement vertex_dict using np.idx
-            vertices = []
-            self.vertex_dict = {}
+            self.vertices = []
+            self.vertex_dict = {} # vertex id to idx in self.vertices
             counter = 0
             for v in raw_vertices:
-                vertices.append(Vertex(v[1], v[2]))
+                self.vertices.append(Vertex(v[1], v[2]))
                 self.vertex_dict[v[0]] = counter
                 counter += 1
-            self.vertices = np.asarray(vertices)
 
-            edges = []
-            self.edge_dict = {}
+            self.edges = []
+            self.edge_dict = {} # sumo edge_id to idx in self.edges
+            self.revert_edge_dict = {} # Edge instance to sumo edge_id
             counter = 0
             for e in raw_edges:
-                edges.append(Edge(self.vertex_dict[e[1]], self.vertex_dict[e[2]]))
+                new_edge = Edge(self.vertex_dict[e[1]], self.vertex_dict[e[2]])
+                self.edges.append(new_edge)
+                self.revert_edge_dict[new_edge] = e[0]
                 self.edge_dict[e[0]] = counter
                 counter += 1
-            self.edges = np.asarray(edges)
 
-            electric_vehicles = []
-            departures = []
-            self.ev_dict = {}
+            self.charging_dict = {} # CS id to idx in self.charging_stations
+            self.station_to_sumo_edge_id = {} # charging_station idx to original sumo edge_id
+            edge_idx_delete = []
+            self.charging_stations = []
             counter = 0
-            for vehicle in raw_departures:
-                electric_vehicles.append(ElectricVehicles(counter, 1, 220, 100))
-                self.ev_dict[vehicle[0]] = counter
+            vtx_counter = len(self.vertices)
```

Comment:
note: I rename the 2 variables both named "counter" and reused it. Please pull my commit on your branch and have a check. (It's minor :p)